### PR TITLE
Marshalling sample

### PR DESCRIPTION
**Pull request by vitek-karas**
Adds a new sample project which shows how to marshal various data types in PInvoke calls.

### CODE REVIEW

File: core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+{
+    stringValueSize = strlen(stringValue) + 1; // + 1 for null terminator
+
+#ifdef WINDOWS
+    // On Windows the returned buffer is going to be freed with CoTaskMemFree, so it needs to be allocated
+    // with CoTaskMemAlloc or CoTaskMemRealloc.
+    char* buffer = (char*)CoTaskMemAlloc(stringValueSize);
+#else
+    // On Linux and Mac the returned buffer is going to be freed with free(), so it needs to be allocated
+    // with malloc/callos/realloc.
```
Comment (jkoritzinsky) [inline core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp]:
```suggestion
    // with malloc/calloc/realloc.
```

File: core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    double result = inValue + *inRef + *inOutRef;
+    *inOutRef = result;
+    *outRef = result;
+    return result;
+}
+
+
+#if !WINDOWS
+typedef struct tagDEC {
+    // Decimal.cs treats the first two shorts as one long
```
Comment (jkoritzinsky) [inline core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp]: Maybe reword this comment as "the runtime" instead of referencing a specific file in CoreCLR?

File: core/interop/pinvoke/marshalling/src/MarshallingSample/MarshallingSampleNative.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    internal static class MarshallingSampleNative
+    {
+        // String marshalling APIs
+        [DllImport(nameof(MarshallingSampleNative), ExactSpelling = true, CharSet = CharSet.Ansi)]
+        public static extern int CountUtf8Characters(string value);
+
+        [DllImport(nameof(MarshallingSampleNative), ExactSpelling = true, CharSet = CharSet.Unicode)]
+        public static extern int CountUtf16Characters(string value);
+
+#if WINDOWS
```
Comment (AaronRobinsonMSFT) [inline core/interop/pinvoke/marshalling/src/MarshallingSample/MarshallingSampleNative.cs]: You might want to comment about CharSet options here. I would at least indicate the default encoding on Windows and what `Ansi` means on non-Windows.

File: core/interop/pinvoke/marshalling/src/MarshallingSample/MarshallingSampleNative.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
+﻿using System;
+using System.Runtime.InteropServices;
+
+namespace MarshallingSample
+{
+    internal static class MarshallingSampleNative
+    {
+        // String marshalling APIs
+        [DllImport(nameof(MarshallingSampleNative), ExactSpelling = true, CharSet = CharSet.Ansi)]
```
Comment (AaronRobinsonMSFT) [inline core/interop/pinvoke/marshalling/src/MarshallingSample/MarshallingSampleNative.cs]: I would remove the `ExactSpelling = true` if possible.

File: core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+{
+    // value is ANSI (on Windows)/ UTF8 (on Linux/Mac) null-terminated string
+    // and can be null.
+    if (value == nullptr)
+    {
+        return -1;
+    }
+    else
+    {
+        return strlen(value);
```
Comment (AaronRobinsonMSFT) [inline core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp]: `CountUtf8Characters()` is a questionable name for this API. `strlen` only counts `char` units. For UTF8, the correct word would be codepoints, but that isn't what `strlen` does. I would instead change to ByteCount or something that doesn't include the expected encoding - i.e. UTF8.

File: core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+    {
+#ifdef WINDOWS
+        return wcslen(value);
+#else
+        return strlen(value);
+#endif
+    }
+}
+
+char* CreateCopyOfString(const char* stringValue, int& stringValueSize)
```
Comment (AaronRobinsonMSFT) [inline core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp]: Mark as `static` since it shouldn't be exported.

File: core/interop/pinvoke/marshalling/readme.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  * Linux/OSX: `g++`
+
+Build and Run
+-------------
+
+1) In order to build and run, all prerequisites must be installed. The following are also required:
+
+    * The C++ compiler (`cl.exe` or `g++`) must be on the path.
+      * On Windows, a [Developer Command Prompt for Visual Studio](https://docs.microsoft.com/cpp/build/building-on-the-command-line#developer_command_prompt_shortcuts) should be used.
+    * The C++ compiler (`cl.exe` or `g++`) and `dotnet` must be the same bitness.
```
Comment (lindexi) [inline core/interop/pinvoke/marshalling/readme.md]: What is [`bitness` ](https://en.wiktionary.org/wiki/bitness) means?

File: core/interop/pinvoke/marshalling/readme.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+      * On Windows the default developer command prompt for VS is using the 32bit compilers but `dotnet` is typically 64bit by default, make sure to pick the "x64 Native Tools Command Prompt for Visual Studio".
```
Comment (lindexi) [inline core/interop/pinvoke/marshalling/readme.md]: My tools name is `x64 Native Tools Command Prompt for VS 2019` 

图片

File: core/interop/pinvoke/marshalling/src/MarshallingSample/EnumMarshalling.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+namespace MarshallingSample
+{
+    class EnumMarshalling
+    {
+        public static void Run()
+        {
+            Console.WriteLine("----- Enum marshalling samples -----");
+
+            // return 2
+            int count = MarshallingSampleNative.CountEnumFlags(MarshallingSampleNative.EnumFlags.A | MarshallingSampleNative.EnumFlags.C);
```
Comment (lindexi) [inline core/interop/pinvoke/marshalling/src/MarshallingSample/EnumMarshalling.cs]: This line is too long. The number of the characters is 126. Maybe we can introduce variable.

         var enumFlags =  MarshallingSampleNative.EnumFlags.A 
                          | MarshallingSampleNative.EnumFlags.C;
         int count = MarshallingSampleNative.CountEnumFlags(enumFlags);

File: core/interop/pinvoke/marshalling/src/MarshallingSample/MarshallingSampleNative.cs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        public static extern int GetStringIntoCalleeAllocatedBuffer(out string value);
+
+        [DllImport(nameof(MarshallingSampleNative), ExactSpelling = true, CharSet = CharSet.Ansi)]
+        public static extern string ReturnStringIntoCalleeAllocatedBuffer();
+
+        [DllImport(nameof(MarshallingSampleNative), ExactSpelling = true, CharSet = CharSet.Ansi)]
+        public static extern int GetStringIntoCallerAllocatedBuffer([In, Out, MarshalAs(UnmanagedType.LPArray)] char[] buffer, ref int bufferSize);
+
+
+        // Int32 marshalling APIs
```
Comment (lindexi) [inline core/interop/pinvoke/marshalling/src/MarshallingSample/MarshallingSampleNative.cs]: Maybe we can replace to region.

File: core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+        return nullptr;
+    }
+
+    memcpy(buffer, stringValue, stringValueSize);
+    return buffer;
+}
+
+extern "C" DLL_EXPORT int STDMETHODCALLTYPE GetStringIntoCalleeAllocatedBuffer(char ** value)
+{
+    const char* stringValue = "Native string value - callee allocated";
```
Comment (lindexi) [inline core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp]: `callee` maybe is `caller`

File: core/interop/pinvoke/marshalling/src/MarshallingSample/MarshallingSampleNative.cs (hunk 4 above)
Comment (vitek-karas) [reply core/interop/pinvoke/marshalling/src/MarshallingSample/MarshallingSampleNative.cs]: I followed the recommendation here: https://docs.microsoft.com/en-us/dotnet/standard/native-interop/best-practices#dllimport-attribute-settings
It says that the recommended value is true (which is not the default) because it is slightly faster.

I'm fine removing it, but if it's our recommendation...

File: core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp (hunk 2 above)
Comment (vitek-karas) [reply core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp]: I deleted the comment - this is sort of a niche thing anyway.

File: core/interop/pinvoke/marshalling/readme.md (hunk 7 above)
Comment (vitek-karas) [reply core/interop/pinvoke/marshalling/readme.md]: 32bit versus 64bit in this context. I added some more explanation to make it clearer.

File: core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp (hunk 11 above)
Comment (vitek-karas) [reply core/interop/pinvoke/marshalling/src/MarshallingSampleNative/MarshallingSampleNative.cpp]: `callee` is correct - the string buffer is allocated by the native function.

File: core/interop/pinvoke/marshalling/readme.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+  * Linux/OSX: `g++`
+
+Build and Run
+-------------
+
+1) In order to build and run, all prerequisites must be installed. The following are also required:
+
+    * The C++ compiler (`cl.exe` or `g++`) must be on the path.
+      * On Windows, a [Developer Command Prompt for Visual Studio](https://docs.microsoft.com/cpp/build/building-on-the-command-line#developer_command_prompt_shortcuts) should be used.
+    * The C++ compiler (`cl.exe` or `g++`) and `dotnet` must be the same bitness (32bit versus 64bit).
```
Comment (rpetrusha) [inline core/interop/pinvoke/marshalling/readme.md]:
```suggestion
    * The C++ compiler (`cl.exe` or `g++`) and `dotnet` must be the same bitness (32-bit versus 64-bit).
```

File: core/interop/pinvoke/marshalling/readme.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+      * On Windows the default developer command prompt for VS is using the 32bit compilers but `dotnet` is typically 64bit by default, make sure to pick the "x64 Native Tools Command Prompt for VS 2019" (or 2017).
```
Comment (rpetrusha) [inline core/interop/pinvoke/marshalling/readme.md]:
```suggestion
      * On Windows, the default developer command prompt for VS uses the 32-bit compilers, but `dotnet` is typically 64-bit by default. Make sure to select the "x64 Native Tools Command Prompt for VS 2019" (or 2017).
```

File: core/interop/pinvoke/marshalling/readme.md [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+1) Navigate to the root directory and run `dotnet build`
+
+1) Run the samples
+
+    * Either use `dotnet run` (which will build and run at the same time)
```
Comment (rpetrusha) [inline core/interop/pinvoke/marshalling/readme.md]:
```suggestion
    * Use `dotnet run` (which will build and run at the same time).
```

File: core/interop/pinvoke/marshalling/readme.md [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+1) In order to build and run, all prerequisites must be installed. The following are also required:
+
+    * The C++ compiler (`cl.exe` or `g++`) must be on the path.
+      * On Windows, a [Developer Command Prompt for Visual Studio](https://docs.microsoft.com/cpp/build/building-on-the-command-line#developer_command_prompt_shortcuts) should be used.
+    * The C++ compiler (`cl.exe` or `g++`) and `dotnet` must be the same bitness (32bit versus 64bit).
+      * On Windows the default developer command prompt for VS is using the 32bit compilers but `dotnet` is typically 64bit by default, make sure to pick the "x64 Native Tools Command Prompt for VS 2019" (or 2017).
+
+1) Navigate to the root directory and run `dotnet build`
+
+1) Run the samples
```
Comment (rpetrusha) [inline core/interop/pinvoke/marshalling/readme.md]:
```suggestion
1) Run the samples. Do one of the following:
```

File: core/interop/pinvoke/marshalling/readme.md [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+    * Either use `dotnet run` (which will build and run at the same time)
+    * Or the executable will be in `bin` under a subdirectory for the configuration (`Debug` is the default).
```
Comment (rpetrusha) [inline core/interop/pinvoke/marshalling/readme.md]:
```suggestion
    * Use `dotnet build` to build the executable. The executable will be in `bin` under a subdirectory for the configuration (`Debug` is the default).
```
```suggestion
    * Or the executable will be in `bin` under a subdirectory for the configuration (`Debug` is the default).
```

File: core/interop/pinvoke/marshalling/readme.md [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+        * Windows: `bin\Debug\MarshallingSample.exe`
+        * Non-Windows: `bin/Debug/MarshallingSample`
+
+Note: The way the sample is built is relatively complicated. The goal is that it's possible to build and run the sample with simple `dotnet run` with minimal requirements on pre-installed tools. Typically real world projects which have both managed and native components will use different build systems for each, for example msbuild/dotnet for managed and CMake for native.
```
Comment (rpetrusha) [inline core/interop/pinvoke/marshalling/readme.md]:
```suggestion
Note: The way the sample is built is relatively complicated. The goal is that it's possible to build and run the sample with simple `dotnet run` with minimal requirements on pre-installed tools. Typically real-world projects which have both managed and native components will use different build systems for each; for example msbuild/dotnet for managed and CMake for native.
```

File: core/interop/pinvoke/marshalling/readme.md [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+Visual Studio support
+---------------------
+
+The `src\MarshallingSample.sln` can be used to open the sample in Visual Studio 2019. In order to be able to build from Visual Studio though it has to be started from the correct developer environment. From the developer environment console start it with `devenv src\MarshallingSample.sln`. With that the solution can be built. To run it set the start project to `MarshallingSample`.
```
Comment (rpetrusha) [inline core/interop/pinvoke/marshalling/readme.md]:
```suggestion
The `src\MarshallingSample.sln` can be used to open the sample in Visual Studio 2019. In order to be able to build from Visual Studio, though, it has to be started from the correct developer environment. From the developer environment console, start it with `devenv src\MarshallingSample.sln`. With that, the solution can be built. To run it set the start project to `MarshallingSample`.
```

File: core/interop/pinvoke/marshalling/src/MarshallingSample/MarshallingSampleNative.cs (hunk 4 above)
Comment (AaronRobinsonMSFT) [reply core/interop/pinvoke/marshalling/src/MarshallingSample/MarshallingSampleNative.cs]: Boo. Okay. Feel free to leave it. Thanks for keeping us honest.

**Comment by rpetrusha** [comment]
Thanks for making the additional changes, @vitek-karas. I'll merge your PR now.
